### DISCUSSION TITLE
Leagues V2

### DISCUSSION THREAD

**Discussion by paulgosnell** [fitlink-app/fitlink-monorepo#564] [Ideas]
As we scale and leagues get busier and the core of our app, some feature and functionality improvements are needed. This is an open discussion to share anything that'll help...

1. Users should only be able to join leagues that match their rank
> See ticket #279 
2. Limit to 100(?) competitors otherwise, the BFIT distributed will be too small 
3. Add waitlist to full leagues so as users leave or get kicked out, new competitors can instantly join
4. Kick out inactive competitors. As discussed with @Jordan-Hall , currently, there seems to be some mechanic that doesn't include a competitor if they haven't competed in the previous round. So if the league resets weekly and they are inactive for the previous week they won't be included in the newly created leaderboard but are still a league member. We need to improve this and remove them from the league altogether but at this same time, notify them that if they don't get active then they will get dropped. 
> See ticket #435 
5. Localised leagues so you can join a league in your timezone. This is because BFIT runs out so if your timezone is in the east you'll have less time than users in the west. 
6. Show points in brackets before BFIT to give some context
> Easy one for @KirillRodichev to implement - ticket created #566 
7. Show up or down chevron against competitor's position so everyone sees who's climbing or dropping 
> I suggest we don’t rely on updates, we should rely on the last time the user checked the leaderboard. In this case we won’t be comparing all leaderboard changes. Will only compare last state when user saw it and the current one
8. Better count-down visualisation as it just counts down in days but on last day we should switch to hours and then minutes to help create a sense of time running out so get out and get active!
9. Better notifications so users know if they are leading, have dropped, have won, will get dropped, time is running out, league reset, friend has joined, invite accepted (some of these live already)
10. Side chat so users can chat about the league, challenge each other, support each other, and general banter to keep people motivated 
11. Link the activities to the leaderboard entries so other competitors can see what each other has been doing in terms of activities.
12. Link the activity pictures to the leaderboard so visually looks good and allow others to link and comment

**Comment by Jordan-Hall** [comment]
Number 5 can be done today with little effort

**Reply by paulgosnell** [reply]
> Number 5 can be done today with little effort

This is how I see 5 (and 6) from a design PoV

![LeagueV2](https://user-images.githubusercontent.com/615605/224557502-02bfc128-cd86-4f6b-9842-ab029f5c5632.png)

**Reply by Jordan-Hall** [reply]
If i get a chance today i'll add it in. Otherwise @KirillRodichev  should be able to quickly add it

**Reply by KirillRodichev** [reply]
Yes, it’s an easy thing

**Reply by Jordan-Hall** [reply]
Do you want to handle it? You will know the colours are better than i will

**Reply by KirillRodichev** [reply]
Yes

**Comment by KirillRodichev** [comment]
How will we track ups and downs for 6? I see some difficulties, but let me know how you see it, so I ask the questions after @paulgosnell

**Reply by paulgosnell** [reply]
> How will we track ups and downs for 6? I see some difficulties, but let me know how you see it, so I ask the questions after @paulgosnell 

I don't know technically how to do this but I guess as the leaderboards update we should keep a record of their current and previous position, then as the leaderboard updates we can calc if their position is higher or lower than the previous position and show the relevant up or down or no change icon. 

As a nice to have, when the leaderboard UI loads, showing the leaderboard entries actually move would be nice... as gives a feel of real-time'ness! But don't want to waste too much time on this like this. Maybe actually a task later is to look at all app animations and do something then.

**Reply by Jordan-Hall** [reply]
@KirillRodichev If you have to fetch the API on the leaderboard i guess you can compare the ordering. That be the "quickest" way of doing things.

**Reply by KirillRodichev** [reply]
What in this case a leaderboard update? If an update would be each position change then we might have the next situation:
- the user doesn’t open the app
- While the app is closed he gains one position and then loses it (so he’s on the same place)
- But compared to his previous position he lost one position
- User opens the app and sees that he lost a position but he actually didn’t move


I suggest we don’t rely on updates, we should rely on the last time the user checked the leaderboard. In this case we won’t be comparing all leaderboard changes. Will only compare last state when user saw it and the current one

**Reply by KirillRodichev** [reply]
@Jordan-Hall, yes. I know how to do it from technical POV, I need to understand the business POV

**Reply by Jordan-Hall** [reply]
Reason for asking because i wasn't sure if I should just track the changes in the backend. Especially saying we know the minute it changes

**Reply by KirillRodichev** [reply]
Oh, ok, I don’t think backend will be changed in order for this to work

**Reply by paulgosnell** [reply]
> I suggest we don’t rely on updates, we should rely on the last time the user checked the leaderboard. In this case we won’t be comparing all leaderboard changes. Will only compare last state when user saw it and the current one

yes @KirillRodichev this sounds good to me 👍
